tools/crm_mon.c

```diff
@@ -1238,9 +1335,15 @@ one_shot(void)
 {
     int rc;
 
-    fencing_connect();
+    rc = pacemakerd_status();
+
+    if (rc == pcmk_rc_ok) {
+        fencing_connect();
+        rc = cib_connect(FALSE);
+    } else {
+        out->err(out, "CIB & fencer sub-daemons not running");
```

Comment:
and here I think we can just let it fall to handle_connection_failures(), which will print an error

Reply:
Hmm ... similar here ... previously cib_connect would have shown specific error-output before the printout of handle_connection_failures.

Reply:
I'd rather just one message, especially since in curses mode it'll be immediately overwritten.

Reply:
Hmm ... think it is worth adding a return-value for one or the other interesting pacemakerd-running-state?

Reply:
Looking at it again, I'm not sure. At the various connection points:

* main() loops for ENOTCONN, and calls handle_connection_failures() for anything else besides pcmk_rc_ok. I think here, what you're currently doing makes sense, so it loops to reconnect.
* one_shot() treats ENOTCONN like any other error, and handle_connection_failures() will output "Error: cluster is not available on this node". Here, richer codes/messages might make sense. Maybe they shouldn't even be errors, e.g. "Pacemaker daemons starting" is more a useful status indicator than an error.
* detect_user_input() also calls handle_connection_failures() for anything but pcmk_rc_ok, which means it would exit. I'm thinking it should probably behave more like main(), i.e. exit only for serious errors and try to reconnect for ENOTCONN.
* reconnect_after_timeout() doesn't call handle_connection_failures() at all. I wonder whether we should treat ENOTCONN as OK and abort on other errors.

Reply:
* in case of one_shot() what the user wants is an immediate status. So if this can't be delivered atm I'd consider this a legitimate error. As implemented atm there will anyway be some additional info about the state of pacemakerd. I'm a bit hesitant changing the interface in a way that we might have significant delays if we wait for a pacemakerd-state-change or something. And returning with a non-error return-code in case we haven't print out a status read from the cib would be problematic for automated calls as e.g. what they get back is probably no proper xml or doesn't fit what they expect otherwise.
* detect_user_input() I guess already does a bit too much already. I'd leave the reconnect-activities to the reconnect-timer and just call refresh_after_event() here - maybe add an enforce parameter. That takes care of the update-timer and reduces the number of paths how to get to the refresh-code.
* guess once we got past the initial loop in main nothing is gonna be that fatal that we shouldn't keep retrying via reconnect_after_timeout(). Maybe here we should as well not call mon_refresh_display() directly but rather go with refresh_after_event() - enforced as above. Btw. handling of the timer is looking a little strange here: first we remove the timer via g_source_remove and forget the reference to the timer then the comment say we would get called again and we return FALSE. Guess we should always return TRUE (with re-engaging the timer in case reconnecting wasn't successful as already done).

Reply:
guess we want "Waiting for CIB ..." just in case of console + curses

Reply:
>     * detect_user_input() I guess already does a bit too much already. I'd leave the reconnect-activities to the reconnect-timer and just call refresh_after_event() here - maybe add an enforce parameter. That takes care of the update-timer and reduces the number of paths how to get to the refresh-code.

You could probably do that with mainloop_set_trigger or mon_trigger_refresh.

>     * guess once we got past the initial loop in main nothing is gonna be that fatal that we shouldn't keep retrying via reconnect_after_timeout(). Maybe here we should as well not call mon_refresh_display() directly but rather go with refresh_after_event() - enforced as above. Btw. handling of the timer is looking a little strange here: first we remove the timer via g_source_remove and forget the reference to the timer then the comment say we would get called again and we return FALSE. Guess we should always return TRUE (with re-engaging the timer in case reconnecting wasn't successful as already done).

The comment is wrong.  Or at least, the "...and reinstall ourselves" part refers to the end of the function.  I've already forgotten why this function is the way it is, but the return codes are important to making things work when connections go away while crm_mon is running.

Just looking at it again, I feel like the g_source_remove and g_timeout_add could just go away.  As long as reconnect_after_timeout returns TRUE, it will keep getting called.  On the other hand, I wonder if it's possible for multiple reconnect_timers to get set up somehow, and that's why we do the remove thing - to cancel a previous timer.

Reply:
should we have refresh_after_event omit triggering the update and stop the refresh_timer when reconnect_timer is engaged? Only interesting thing might be changing the display-settings and have the status from the last cib-update prior to disconnection shown. But after cib-disconnect event the screen is anyway cleared and we see the Reconnect... message.

Reply:
Now that we are pinging pacemakerd for all daemons started race that cib can connect but fencer can't isn't very likely anymore.
But if it happens - as fencer connection failing is assumed non fatal - it will never recover.
As long as we triggered a reconnection in detect_user_input typing anything would have triggered fencer reconnection - not clean but still a possibility ...
Anything speaking against trying a fencer reconnection with the refreshes?

Reply:
>     * in case of one_shot() what the user wants is an immediate status. So if this can't be delivered atm I'd consider this a legitimate error. As implemented atm there will anyway be some additional info about the state of pacemakerd. I'm a bit 

That makes sense

> hesitant changing the interface in a way that we might have significant delays if we wait for a pacemakerd-state-change or something. And returning with a non-error return-code in case we haven't print out a status read from the cib would be problematic for automated calls as e.g. what they get back is probably no proper xml or doesn't fit what they expect otherwise.

I definitely don't want to reconnect or anything like that in one-shot. It's a good question what users would prefer when the cluster is known to be starting/stopping and full information isn't available. It's reasonable to stick with error since that's how it is now, but it seems to me that it's legitimate state information rather than an error.

>     * detect_user_input() I guess already does a bit too much already. I'd leave the reconnect-activities to the reconnect-timer and just call refresh_after_event() here - maybe add an enforce parameter. That takes care of the update-timer and reduces the number of paths how to get to the refresh-code.

Something like that makes sense, we're definitely at the point where the connection/display code is duplicated a bit much.

>     * guess once we got past the initial loop in main nothing is gonna be that fatal that we shouldn't keep retrying via reconnect_after_timeout(). Maybe here we should as well not call mon_refresh_display() directly but rather go with refresh_after_event() - enforced as above.

Yep, once we're in the main loop, using the trigger is better than calling mon_refresh_display() directly. one_shot() doesn't use the main loop so we need to make sure that path is kept separate.

> Btw. handling of the timer is looking a little strange here: first we remove the timer via g_source_remove and forget the reference to the timer then the comment say we would get called again and we return FALSE. Guess we should always return TRUE (with re-engaging the timer in case reconnecting wasn't successful as already done).

I did think that looked weird. I wouldn't reset the timer explicitly, just return TRUE or FALSE as appropriate. BTW, now that we're requiring a higher glib version, we can use G_SOURCE_CONTINUE and G_SOURCE_REMOVE instead of TRUE and FALSE, for readability.

Reply:
> should we have refresh_after_event omit triggering the update and stop the refresh_timer when reconnect_timer is engaged? Only interesting thing might be changing the display-settings and have the status from the last cib-update prior to disconnection shown. But after cib-disconnect event the screen is anyway cleared and we see the Reconnect... message.

I wouldn't want to show information we know is outdated.

Reply:
> Now that we are pinging pacemakerd for all daemons started race that cib can connect but fencer can't isn't very likely anymore.
> But if it happens - as fencer connection failing is assumed non fatal - it will never recover.
> As long as we triggered a reconnection in detect_user_input typing anything would have triggered fencer reconnection - not clean but still a possibility ...
> Anything speaking against trying a fencer reconnection with the refreshes?

Makes sense to me. I also don't see a reason to tie the CIB and fencer connections together, reconnecting just what needs reconnecting should be fine.

Reply:
> > should we have refresh_after_event omit triggering the update and stop the refresh_timer when reconnect_timer is engaged? Only interesting thing might be changing the display-settings and have the status from the last cib-update prior to disconnection shown. But after cib-disconnect event the screen is anyway cleared and we see the Reconnect... message.
> 
> I wouldn't want to show information we know is outdated.

Me neither. Just wanted to assure we aren't missing anything if we suppress display updates while reconnecting.

Reply:
> 
> I definitely don't want to reconnect or anything like that in one-shot. It's a good question what users would prefer when the cluster is known to be starting/stopping and full information isn't available. It's reasonable to stick with error since that's how it is now, but it seems to me that it's legitimate state information rather than an error.
> 

Just assuring we are on the same page ...
Then I'd delay the switch to a state information without error return for the next overhaul - till it is clear what e.g. pcs or other high-level tooling would do with it

> 
> Yep, once we're in the main loop, using the trigger is better than calling mon_refresh_display() directly. one_shot() doesn't use the main loop so we need to make sure that path is kept separate.

Of course - kept the call in one_shot as is.

> 
> I did think that looked weird. I wouldn't reset the timer explicitly, just return TRUE or FALSE as appropriate. BTW, now that we're requiring a higher glib version, we can use G_SOURCE_CONTINUE and G_SOURCE_REMOVE instead of TRUE and FALSE, for readability.

It is probably cleaner to have the timer removed implicitly compared to doing that in the handler.
Probably the reason why it did return TRUE so that glib wouldn't remove the timer that had already been removed.
Adding a new timer right at the end probably still makes sense because it assures that it starts ticking just then (the formulation in the glib reference sounds as if it would do the same - recalculate after each call to the timeout function - but I think it doesn't).

Reply:
> > Now that we are pinging pacemakerd for all daemons started race that cib can connect but fencer can't isn't very likely anymore.
> > But if it happens - as fencer connection failing is assumed non fatal - it will never recover.
> > As long as we triggered a reconnection in detect_user_input typing anything would have triggered fencer reconnection - not clean but still a possibility ...
> > Anything speaking against trying a fencer reconnection with the refreshes?
> 
> Makes sense to me. I also don't see a reason to tie the CIB and fencer connections together, reconnecting just what needs reconnecting should be fine.

Loss of fencer-connection tears down the cib-connection as well and the other way round. I had implemented it like that back then as it keeps the number of possible patterns smaller and the code simpler. But that was just fully true as long as we were handling the fencer-connection as fatal. Should we change that as well now or postpone?

Reply:
Either way is fine